text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1015 @@
+  @tracked lastName = 'Dale';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+This also allows us to opt out of tracking entirely, like if we know for
+instance that a given property is constant and will never change. In general,
+the idea is that _mutable_ properties should be marked as tracked, and
+_immutable_ properties should not.
+
+### Reducing Memory Consumption
+
+By default, computed properties cache their values. This is great when a
+computed property has to perform expensive work to produce its value, and that
+value gets used over and over again.
+
+But checking, populating, and invalidating this cache comes with its own
+overhead. Modern JavaScript VMs can produce highly optimized code, and in many
+cases the overhead of caching is greater than the cost of simply recomputing the
+value.
+
+Worse, cached computed property values cannot be freed by the garbage collector
+until the entire object is freed. Many computed properties are accessed only
+once, but because they cache by default, they take up valuable space on the heap
+for no benefit.
+
+For example, imagine this component that checks whether the `files` property is
+supported in input elements:
+
+```js
+import Component from '@ember/component';
+import { computed } from '@ember/object';
+
+export default Component.extend({
+  inputElement: computed(function() {
+    return document.createElement('input');
+  }),
+
+  supportsFiles: computed('inputElement', function() {
+    return 'files' in this.inputElement;
+  }),
+
+  didInsertElement() {
+    if (this.supportsFiles) {
+      // do something
+    } else {
+      // do something else
+    }
+  },
+});
+```
+
+This component would create and retain an `HTMLInputElement` DOM node for the
+lifetime of the component, even though all we really want to cache is the
+Boolean value of whether the browser supports the `files` attribute.
+
+Particularly on inexpensive mobile devices, where RAM is limited and often slow, we should
+be more conservative about our memory consumption. Tracked properties switch
+from an opt-out caching model to opt-in, allowing developers to err on the side
+of reduced memory usage, but easily enabling caching (a.k.a. memoization) if a
+property shows up as a bottleneck during profiling.
+
+## Prior Art
+
+Tracked properties were first implemented in [Glimmer.js](https://glimmerjs.com/),
+and were recently polyfilled with clever usage of `notifyPropertyChange` by
+the [sparkles-components](https://github.com/rwjblue/sparkles-component/) addon.
+These initial implementations inform the design in this RFC, but differ from it
+in some key ways. For instance, both Sparkles's and early versions of Glimmer's
+`@tracked` did not have an autotracking stack, and instead relied on explicit
+dependency keys. After benchmarking showed that autotracking was a viable
+strategy, the API for `@tracked` was updated to what is proposed here.
+
+## Detailed Design
```

Comment:
@pzuraq what would be the intended API for a proxy use case?  For example if today you have an `ObjectProxy`, or some class that implements `unknownProperty` and `setUnknownProperty`, and this object's properties are accessed directly in templates, what's the intended mechanism for communicating that these properties are tracked?

Reply:
This has been addressed now directly, but it should continue to work through using `get` and `set` (which proxies require you to do). When using `get`, tags will be created for any properties accessed, and `set` will invalidate those tags.

Internally, classes that implement `unknownProperty` will have to either use `get` or CPs to make sure that whatever they are proxying to also updates in this way, or that `notifyPropertyChange` is called for the proxied property somehow at some point.